### PR TITLE
Port stateless timer to exp manager

### PR DESCRIPTION
**Pull request by MaximumEntropy**
Signed-off-by: MaximumEntropy <sandeep.subramanian.1@umontreal.ca>

# What does this PR do ?

Allows configuring a stateless timer via `exp_manager.max_time=03:55:00`.

**Collection**: All

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nemo/utils/exp_manager.py [hunk 1]
```diff
@@ -402,6 +402,12 @@ def exp_manager(trainer: 'pytorch_lightning.Trainer', cfg: Optional[Union[DictCo
         # extend training loop to skip initial validation when resuming from checkpoint
         configure_no_restart_validation_training_loop(trainer)
 
+    # Setup a stateless timer for use on clusters.
+    if cfg.max_time is not None:
```
Comment (titu1994) [inline nemo/utils/exp_manager.py:406]: Need to add this key to the exp manager config.

File: nemo/utils/exp_manager.py [hunk 2]
```diff
@@ -402,6 +402,12 @@ def exp_manager(trainer: 'pytorch_lightning.Trainer', cfg: Optional[Union[DictCo
...
+        if trainer.max_time is not None:
+            raise ValueError(f"trainer.max_time is not None, Cannot set both trainer.max_time and exp_manager.max_time")
+        trainer.callbacks.append(StatelessTimer(cfg.max_time))
```
Comment (titu1994) [inline nemo/utils/exp_manager.py:411]: And update trainer.max_time

**Review by titu1994 — CHANGES_REQUESTED** [review]

File: nemo/utils/exp_manager.py (hunk 1 above)
Comment (MaximumEntropy) [reply nemo/utils/exp_manager.py:406]: Done!

File: nemo/utils/exp_manager.py (hunk 2 above)
Comment (MaximumEntropy) [reply nemo/utils/exp_manager.py]: Done!

**Review by titu1994 — DISMISSED** [review]

**Comment by titu1994** [comment]
Is this ready to merge ? @MaximumEntropy

**Review by titu1994 — APPROVED** [review]

**Comment by titu1994** [comment]
It looks good!

**Comment by MaximumEntropy** [comment]
@titu1994 I've made some changes specifically around how I handle a user setting both `trainer.max_time` as well as `exp_manager.max_time_per_run`. PTL does not set a `max_time` variable for me to override and PTL timers get set up before exp_manager and so I cannot intervene. So I'm just throwing a warning and overriding the PTL timer with the stateless one if a user sets both. (details in the comments in the code).

**Comment by titu1994** [comment]
Yes I saw that. An unfortunate side effect but we could ask for an extensible Timer impl in PTL. This is sufficient for now

**Comment by MaximumEntropy** [comment]
Okay I tested this with Megatron runs and it works - turns out I already had a unit test for this and just change the config of that unit test to use exp_manager.max_time_per_run instead of `trainer.max_time`.

**Pull request merged**